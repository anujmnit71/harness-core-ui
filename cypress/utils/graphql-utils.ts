export const hasOperationName: (req: any, operationName: string) => boolean = (req, operationName) => {
  const { body } = req
  // eslint-disable-next-line no-prototype-builtins
  return body.hasOwnProperty('operationName') && body.operationName === operationName
}

export const aliasQuery: (req: any, operationName: string) => void = (req, operationName) => {
  if (hasOperationName(req, operationName)) {
    req.alias = `gql${operationName}Query`
  }
}
