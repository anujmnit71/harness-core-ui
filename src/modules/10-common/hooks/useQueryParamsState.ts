/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { isEqual } from 'lodash-es'
import { useQueryParams } from './useQueryParams'
import { useUpdateQueryParams } from './useUpdateQueryParams'

/**
 * Works like useState except it reads and sets state in query params
 */

interface UseQueryParamsStateOptions<T> {
  serializer: (data: T) => string
  deserializer: (data: string) => T
}

export function useQueryParamsState<T = unknown>(
  name: string,
  defaultValue: T,
  options: UseQueryParamsStateOptions<T> = {
    serializer: JSON.stringify,
    deserializer: JSON.parse
  }
): [T, (newState: T) => void] {
  const { updateQueryParams } = useUpdateQueryParams()

  const { state } = useQueryParams<{ state: T }>({
    processQueryParams(params) {
      const stateFromQuery = params[name]
      return {
        state: stateFromQuery ? options.deserializer(stateFromQuery) : defaultValue
      }
    }
  })

  const setState: (newState: T) => void = newState => {
    !isEqual(state, newState) && updateQueryParams({ [name]: options.serializer(newState) })
  }

  return [state, setState]
}
