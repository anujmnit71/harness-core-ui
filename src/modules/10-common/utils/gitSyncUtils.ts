/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import type { GitSyncConfig } from 'services/cd-ng'
import { folderPathName, yamlFileExtension } from './StringUtils'

export const getRepoDetailsByIndentifier = (identifier: string | undefined, repos: GitSyncConfig[]) => {
  return repos.find((repo: GitSyncConfig) => repo.identifier === identifier)
}

export const validateFilePath = (filePath: string): boolean => {
  if (!filePath.endsWith(yamlFileExtension)) {
    return false
  } else {
    const fullPath = filePath.slice(0, filePath.length - yamlFileExtension.length)
    const subPaths = fullPath.split('/')
    return subPaths.every((folderName: string) => {
      return folderName.length && folderName.match(folderPathName)?.length
    })
  }
}
