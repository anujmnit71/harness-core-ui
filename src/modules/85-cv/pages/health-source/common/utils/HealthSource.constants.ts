/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

export const HealthSourceFieldNames = {
  MANUAL_QUERY: 'isManualQuery',
  IDENTIFIER: 'identifier',
  METRIC_NAME: 'metricName',
  METRIC_DATA: 'metricData',
  GROUP_NAME: 'groupName',
  CONTINUOUS_VERIFICATION: 'continuousVerification',
  HEALTH_SCORE: 'healthScore',
  SLI: 'sli',
  RISK_CATEGORY: 'riskCategory',
  HIGHER_BASELINE_DEVIATION: 'higherBaselineDeviation',
  LOWER_BASELINE_DEVIATION: 'lowerBaselineDeviation'
}
