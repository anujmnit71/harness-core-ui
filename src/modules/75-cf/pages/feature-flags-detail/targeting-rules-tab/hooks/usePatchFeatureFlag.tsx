/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { useParams } from 'react-router-dom'
import { useToaster } from '@harness/uicore'
import { get } from 'lodash-es'
import patch from '@cf/utils/instructions'
import { FeatureState, PatchFeatureQueryParams, usePatchFeature } from 'services/cf'
import useActiveEnvironment from '@cf/hooks/useActiveEnvironment'
import { showToaster } from '@cf/utils/CFUtils'
import { useStrings } from 'framework/strings'
import type { TargetingRulesFormValues } from '../TargetingRulesTab'

interface UsePatchFeatureFlagProps {
  featureFlagIdentifier: string
  initialValues: TargetingRulesFormValues
}

interface UsePatchFeatureFlagReturn {
  saveChanges: (values: TargetingRulesFormValues, onSuccess: () => void) => void
  loading: boolean
}

const usePatchFeatureFlag = ({
  featureFlagIdentifier,
  initialValues
}: UsePatchFeatureFlagProps): UsePatchFeatureFlagReturn => {
  const { projectIdentifier, orgIdentifier, accountId: accountIdentifier } = useParams<Record<string, string>>()
  const { activeEnvironment: environmentIdentifier } = useActiveEnvironment()
  const { showError } = useToaster()
  const { getString } = useStrings()

  const { mutate: patchFeature, loading } = usePatchFeature({
    identifier: featureFlagIdentifier as string,
    queryParams: {
      projectIdentifier,
      environmentIdentifier,
      accountIdentifier,
      orgIdentifier
    } as PatchFeatureQueryParams
  })

  const saveChanges = (values: TargetingRulesFormValues, onSuccess: () => void): void => {
    if (values.state !== initialValues.state) {
      patch.feature.addInstruction(patch.creators.setFeatureFlagState(values?.state as FeatureState))
    }

    try {
      patch.feature.onPatchAvailable(async data => {
        await patchFeature(data)

        showToaster(getString('cf.messages.flagUpdated'))

        onSuccess()
      })

      patch.feature.reset()
    } catch (error: any) {
      showError(get(error, 'data.message', error?.message), 0)
    }
  }
  return {
    saveChanges,
    loading
  }
}

export default usePatchFeatureFlag
