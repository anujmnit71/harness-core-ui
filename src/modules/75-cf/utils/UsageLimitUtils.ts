/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import type { ReactNode } from 'react'
import { BannerType } from '@common/layouts/Constants'
import type { CheckFeatureReturn } from 'framework/featureStore/featureStoreUtil'
import { formatToCompactNumber } from '@cf/utils/CFUtils'
import type { UseStringsReturn } from 'framework/strings'

export const getBannerText = (
  getString: UseStringsReturn['getString'],
  monthlyActiveUsers: CheckFeatureReturn,
  additionalLicenseProps: Record<string, any>
): ReactNode => {
  const { isEnterpriseEdition, isFreeEdition, isTeamEdition } = additionalLicenseProps

  const clientMauUsageCount = Number(monthlyActiveUsers?.featureDetail?.count)
  const clientMauPlanLimit = Number(monthlyActiveUsers?.featureDetail?.limit)

  const clientMauUsagePercentage = Math.trunc((clientMauUsageCount / clientMauPlanLimit) * 100)
  const clientMauPlanLimitFormatted = formatToCompactNumber(clientMauPlanLimit)

  const showInfoBanner = clientMauPlanLimit > 0 && clientMauUsagePercentage >= 90 && clientMauUsagePercentage < 100
  const showWarningBanner = clientMauPlanLimit > 0 && clientMauUsagePercentage >= 100

  let message!: string
  let bannerType!: BannerType

  if (isFreeEdition) {
    if (!showInfoBanner) {
      return {
        message: () =>
          getString('cf.planEnforcement.freePlan.approachingLimit', {
            clientMauUsagePercentage,
            clientMauPlanLimitFormatted
          }),
        bannerType: BannerType.INFO
      }
    }

    if (showWarningBanner) {
      return {
        message: getString('cf.planEnforcement.freePlan.upgradeRequired', {
          clientMauPlanLimitFormatted
        }),
        bannerType: BannerType.LEVEL_UP
      }
    }
  }

  if (isTeamEdition || isEnterpriseEdition) {
    if (showInfoBanner) {
      return {
        message: () =>
          getString('cf.planEnforcement.teamEnterprisePlan.approachingLimit', {
            clientMauUsagePercentage
          }),
        bannerType: BannerType.INFO
      }
    }

    if (showWarningBanner) {
      return {
        message: () => getString('cf.planEnforcement.teamEnterprisePlan.upgradeRequired'),
        bannerType: BannerType.LEVEL_UP
      }
    }
  }

  return {
    message,
    bannerType
  }
}
