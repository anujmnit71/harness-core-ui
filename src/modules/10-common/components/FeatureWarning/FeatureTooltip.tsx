/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React, { ReactElement } from 'react'
import { Color, FontVariation, Layout, Text } from '@harness/uicore'
import { capitalize } from 'lodash-es'
import type { IconName } from '@blueprintjs/core'
import { useStrings } from 'framework/strings'
import { FeatureDescriptor, customFeatureDescriptor } from 'framework/featureStore/FeatureDescriptor'
import { useFeatureRequiredPlans } from '@common/hooks/useFeatures'
import type { FeatureIdentifier } from 'framework/featureStore/FeatureIdentifier'
import type { StringsMap } from 'stringTypes'
import type { CheckFeatureReturn, ModuleType } from 'framework/featureStore/featureStoreUtil'
import { ExplorePlansBtn } from './featureWarningUtil'
import css from './FeatureWarning.module.scss'

interface FeatureTooltipProps {
  features: Map<FeatureIdentifier, CheckFeatureReturn>
  warningMessage?: React.ReactNode
}

const FeatureText = ({ feature, module }: { feature: CheckFeatureReturn; module: ModuleType }): React.ReactElement => {
  const { getString } = useStrings()
  const { count, featureName } = feature.featureDetail || {}
  const requiredPlans = useFeatureRequiredPlans(featureName)
  if (featureName) {
    const featureNameStr = capitalize(featureName.split('_').join(' '))
    const customFeatureDescription = customFeatureDescriptor(featureName, getString, { limits: count })
    const featureDescription = FeatureDescriptor[featureName] ? FeatureDescriptor[featureName] : featureNameStr
    const isCore = module === 'CORE'
    const limitDescription = (
      <Text padding={{ bottom: 'small', left: 'large' }} color={Color.WHITE}>
        {`${getString('common.feature.levelUp.limitMessage.description' as keyof StringsMap, {
          limits: count
        })} `}
        {featureDescription}
        {` ${getString('common.feature.levelUp.limitMessage.thisMonth').toLowerCase()}.`}
      </Text>
    )
    if (!isCore && count) {
      return customFeatureDescription ? (
        <Text padding={{ bottom: 'small', left: 'large' }} color={Color.WHITE}>
          {customFeatureDescription}
        </Text>
      ) : (
        limitDescription
      )
    }
    const requiredPlansStr =
      requiredPlans.length > 0
        ? requiredPlans.join(` ${getString('common.or').toLowerCase()} `)
        : getString('common.upgrade').toLowerCase()
    const upgradeDescription = ` ${getString('common.feature.levelUp.planMessage' as keyof StringsMap, {
      plan: requiredPlansStr
    })}`
    return (
      <Text padding={{ bottom: 'small' }} color={Color.WHITE}>
        {featureDescription}
        {upgradeDescription}
      </Text>
    )
  }
  return <></>
}

const FeatureTooltip = ({ features, warningMessage }: FeatureTooltipProps): ReactElement => {
  const { getString } = useStrings()
  const firstDisabledFeature = [...features.keys()].find(key => features.get(key)?.enabled === false)

  function getGroupDisabledFeatures(): Map<ModuleType, CheckFeatureReturn[]> {
    return [...features.values()].reduce((group, feature) => {
      const module = feature.featureDetail?.moduleType
      const featureDisabled = !feature.enabled
      if (featureDisabled) {
        if (module && ['CD', 'CI', 'CE', 'CF'].includes(module)) {
          group.set(module, [...(group.get(module) || []), feature])
        } else {
          group.set('CORE', [...(group.get('CORE') || []), feature])
        }
      }
      return group
    }, new Map<ModuleType, CheckFeatureReturn[]>())
  }

  function getGroupedTooltips(): React.ReactNode {
    const groupedDisabledFeatures = getGroupDisabledFeatures()

    function getHeader(module: ModuleType): React.ReactNode | undefined {
      let icon, id
      switch (module) {
        case 'CD':
          icon = 'cd-main'
          id = 'common.purpose.cd.continuous'
          break

        case 'CI':
          icon = 'ci-main'
          id = 'common.purpose.ci.continuous'
          break
        case 'CF':
          icon = 'cf-main'
          id = 'common.purpose.cf.continuous'
          break
        case 'CE':
          icon = 'ce-main'
          id = 'common.purpose.ce.continuous'
          break
      }
      return (
        icon &&
        id && (
          <Text
            icon={icon as IconName}
            color={Color.WHITE}
            iconProps={{ color: Color.WHITE }}
            font={{ weight: 'semi-bold' }}
            padding={{ bottom: 'small' }}
          >
            {getString(id as keyof StringsMap)}
          </Text>
        )
      )
    }

    const haslimits = [...groupedDisabledFeatures.values()].find(value =>
      value.find(feature => feature.featureDetail?.limit !== undefined)
    )

    return (
      <Layout.Vertical spacing="large">
        {haslimits && (
          <Text color={Color.WHITE} padding={{ bottom: 'large' }} rightIcon="flag" className={css.flag}>
            {getString('common.feature.levelUp.description')}
          </Text>
        )}
        {[...groupedDisabledFeatures.keys()].map(key => {
          const disabledFeatures = groupedDisabledFeatures.get(key)
          return (
            <Layout.Vertical key={key} padding={{ bottom: 'large' }}>
              <Layout.Vertical padding={{ bottom: 'small' }}>
                {getHeader(key)}
                {disabledFeatures?.map(feature => (
                  <FeatureText key={feature.featureDetail?.featureName} feature={feature} module={key} />
                ))}
              </Layout.Vertical>
              <ExplorePlansBtn featureName={disabledFeatures?.[0].featureDetail?.featureName} className={css.btn} />
            </Layout.Vertical>
          )
        })}
      </Layout.Vertical>
    )
  }

  function getTooltipBody(): React.ReactNode {
    // return customized warning message
    if (warningMessage) {
      return (
        <>
          <Text font={{ size: 'small' }} color={Color.WHITE} padding={{ bottom: 'large' }}>
            {warningMessage}
          </Text>
          <ExplorePlansBtn featureName={firstDisabledFeature} className={css.btn} />
        </>
      )
    }

    // group by module messages if there are any
    return getGroupedTooltips()
  }

  return (
    <Layout.Vertical padding="medium" className={css.tooltip}>
      <Text
        icon="flash"
        color={Color.ORANGE_800}
        font={{ variation: FontVariation.FORM_MESSAGE_WARNING, weight: 'bold' }}
        iconProps={{ color: Color.ORANGE_800, size: 25 }}
        padding={{ bottom: 'xsmall' }}
      >
        {getString('common.levelUp')}
      </Text>
      {getTooltipBody()}
    </Layout.Vertical>
  )
}

export default FeatureTooltip
