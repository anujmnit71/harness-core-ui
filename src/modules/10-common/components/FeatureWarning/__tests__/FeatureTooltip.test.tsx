/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import { render } from '@testing-library/react'
import { TestWrapper } from '@common/utils/testUtils'
import routes from '@common/RouteDefinitions'
import { accountPathProps } from '@common/utils/routeUtils'
import { FeatureIdentifier } from 'framework/featureStore/FeatureIdentifier'
import { useFeatureModule, useFeatureRequiredPlans } from '@common/hooks/useFeatures'
import type { CheckFeatureReturn } from 'framework/featureStore/featureStoreUtil'
import FeatureTooltip from '../FeatureTooltip'

jest.mock('@common/hooks/useFeatures')
const useFeatureModuleMock = useFeatureModule as jest.MockedFunction<any>
useFeatureModuleMock.mockImplementation(() => {
  return 'CI'
})

const useFeatureRequiredPlansMock = useFeatureRequiredPlans as jest.MockedFunction<any>
useFeatureRequiredPlansMock.mockImplementation(() => {
  return []
})

const features = new Map<FeatureIdentifier, CheckFeatureReturn>()
features.set(FeatureIdentifier.BUILDS, {
  enabled: false,
  featureDetail: {
    featureName: FeatureIdentifier.BUILDS,
    enabled: false,
    moduleType: 'CD',
    limit: 100,
    count: 200
  }
})
features.set(FeatureIdentifier.DEPLOYMENTS, {
  enabled: false,
  featureDetail: {
    featureName: FeatureIdentifier.DEPLOYMENTS,
    enabled: false,
    moduleType: 'CI',
    limit: 100,
    count: 200
  }
})

const warningMesasge = <p>this is warning message</p>

describe('FeatureTooltip', () => {
  test('FeatureTooltip with warning message', () => {
    const { container, getByText } = render(
      <TestWrapper path={routes.toRoleDetails({ ...accountPathProps })} pathParams={{ accountId: 'dummy' }}>
        <FeatureTooltip features={features} warningMessage={warningMesasge} />
      </TestWrapper>
    )
    expect(getByText('this is warning message')).toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })

  test('FeatureTooltip without warning message', () => {
    const { container, queryByText, getByText } = render(
      <TestWrapper path={routes.toRoleDetails({ ...accountPathProps })} pathParams={{ accountId: 'dummy' }}>
        <FeatureTooltip features={features} />
      </TestWrapper>
    )
    expect(queryByText('this is warning message')).not.toBeInTheDocument()
    expect(getByText('common.purpose.cd.continuous')).toBeInTheDocument()
    expect(getByText('common.purpose.ci.continuous')).toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })
})
