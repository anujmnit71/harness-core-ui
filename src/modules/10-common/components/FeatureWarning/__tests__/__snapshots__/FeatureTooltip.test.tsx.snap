// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeatureTooltip FeatureTooltip with warning message 1`] = `
<div>
  <div
    class="Layout--vertical StyledProps--main tooltip StyledProps--padding-medium"
  >
    <p
      class="StyledProps--font StyledProps--main StyledProps--color StyledProps--color-orange800 StyledProps--font-variation-form-message-warning StyledProps--font-weight-bold StyledProps--padding-bottom-xsmall"
      style="display: flex; align-items: center;"
    >
      <span
        class="bp3-icon StyledProps--main Text--icon StyledProps--padding-right-xsmall StyledProps--color StyledProps--color-orange800"
        data-icon="flash"
      >
        <svg
          height="25"
          viewBox="0 0 33 28"
          width="25"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M3.773 17.341L20.885 1.75l-2.297 8.909h8.184L9.66 26.25l2.297-8.909z"
            fill="#ff5310"
            fill-rule="evenodd"
          />
        </svg>
      </span>
      common.levelUp
    </p>
    <p
      class="StyledProps--font StyledProps--main StyledProps--font-size-small StyledProps--color StyledProps--color-white StyledProps--padding-bottom-large"
    >
      <p>
        this is warning message
      </p>
    </p>
    <button
      class="bp3-button Button--button StyledProps--font StyledProps--main btn Button--with-current-color Button--iconOnly Button--variation Button--variation-primary Button--size Button--small"
      type="button"
    >
      <span
        class="bp3-button-text"
      >
        common.explorePlans
      </span>
    </button>
  </div>
</div>
`;

exports[`FeatureTooltip FeatureTooltip without warning message 1`] = `
<div>
  <div
    class="Layout--vertical StyledProps--main tooltip StyledProps--padding-medium"
  >
    <p
      class="StyledProps--font StyledProps--main StyledProps--color StyledProps--color-orange800 StyledProps--font-variation-form-message-warning StyledProps--font-weight-bold StyledProps--padding-bottom-xsmall"
      style="display: flex; align-items: center;"
    >
      <span
        class="bp3-icon StyledProps--main Text--icon StyledProps--padding-right-xsmall StyledProps--color StyledProps--color-orange800"
        data-icon="flash"
      >
        <svg
          height="25"
          viewBox="0 0 33 28"
          width="25"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M3.773 17.341L20.885 1.75l-2.297 8.909h8.184L9.66 26.25l2.297-8.909z"
            fill="#ff5310"
            fill-rule="evenodd"
          />
        </svg>
      </span>
      common.levelUp
    </p>
    <div
      class="Layout--vertical Layout--layout-spacing-large StyledProps--main"
    >
      <p
        class="StyledProps--font StyledProps--main flag StyledProps--color StyledProps--color-white StyledProps--padding-bottom-large"
        style="display: flex; align-items: center;"
      >
        common.feature.levelUp.description
        <span
          class="bp3-icon StyledProps--main Text--icon StyledProps--padding-left-xsmall"
          data-icon="flag"
        >
          <svg
            fill="none"
            height="16"
            viewBox="0 0 32 32"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <clippath
              id="flag_svg__a"
            >
              <path
                d="M0 0h32v32H0z"
              />
            </clippath>
            <g
              clip-path="url(#flag_svg__a)"
            >
              <path
                d="M4.445 32a1.778 1.778 0 01-1.778-1.778V2.667a1.778 1.778 0 013.556 0v27.555c0 .982-.797 1.778-1.778 1.778z"
                fill="#8899a6"
              />
              <path
                d="M4.445.889c-.982 0-1.778.795-1.778 1.778v27.555a.445.445 0 00.889 0V3.924c0-.695.563-1.257 1.257-1.257h1.41c0-.983-.796-1.778-1.778-1.778z"
                fill="#aab8c2"
              />
              <path
                d="M4.445 32a1.778 1.778 0 01-1.778-1.778V2.667a1.778 1.778 0 013.556 0v27.555c0 .982-.797 1.778-1.778 1.778z"
                fill="#8899a6"
              />
              <path
                d="M4.445.889c-.982 0-1.778.795-1.778 1.778v27.555a.445.445 0 00.889 0V3.924c0-.695.563-1.257 1.257-1.257h1.41c0-.983-.796-1.778-1.778-1.778z"
                fill="#aab8c2"
              />
              <path
                d="M28.796 2.74c-1.479-.64-3.055-.957-4.796-.985l-1.223.337-4.11.575-.89-.114-.632.133c-1.928.417-3.76.795-5.59.873l-.889.885-5.333-.888v16.888h5.333l.89 1.778h.068c1.994-.085 3.975-.494 5.896-.908l.257-.054.575-.574 4.833-.63.815.365c1.493.029 2.838.298 4.092.84a.888.888 0 001.241-.817V3.556a.888.888 0 00-.537-.816z"
                fill="#31373d"
              />
              <path
                d="M11.555 3.56c-.212.008-.425.03-.637.03-1.597 0-3.019-.278-4.344-.85a.881.881 0 00-.84.073.886.886 0 00-.401.743v5.333s2.926.889 6.222.889zm6.222 5.329s-3.333.889-6.222.889V16c2.667 0 6.222-.889 6.222-.889zM24 8V1.755c-.086-.002-.166-.012-.252-.012-2.015 0-4.019.393-5.97.81v6.336S19.703 8 24 8zM5.87 21.26c1.333.576 2.747.883 4.284.962h1.401V16c-3.704 0-6.222-.889-6.222-.889v5.333c0 .355.21.676.537.817zm11.908-6.149v6.148c2.056-.444 4.01-.85 5.97-.85.086 0 .166.01.252.012v-6.199c-4 0-6.222.89-6.222.89zM24 14.222c2.378 0 4.284.498 5.333.848V8.807C28.313 8.47 26.461 8 24 8z"
                fill="#e1e8ed"
              />
            </g>
          </svg>
        </span>
      </p>
      <div
        class="Layout--vertical StyledProps--main StyledProps--padding-bottom-large"
      >
        <div
          class="Layout--vertical StyledProps--main StyledProps--padding-bottom-small"
        >
          <p
            class="StyledProps--font StyledProps--main StyledProps--color StyledProps--color-white StyledProps--font-weight-semi-bold StyledProps--padding-bottom-small"
            style="display: flex; align-items: center;"
          >
            <span
              class="bp3-icon StyledProps--main Text--icon StyledProps--padding-right-xsmall StyledProps--color StyledProps--color-white"
              data-icon="cd-main"
            >
              <svg
                height="16"
                viewBox="0 0 32 32"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  clip-rule="evenodd"
                  d="M26.427 7.06A5.073 5.073 0 0129 11.52l-.077 8.934a5.073 5.073 0 01-2.537 4.35l-8.002 4.618a5.073 5.073 0 01-5.035.021l-7.776-4.402A5.073 5.073 0 013 20.583l.077-8.935a5.073 5.073 0 012.537-4.35l8.002-4.619a5.073 5.073 0 015.035-.02l7.776 4.402zm-2.83 6.938c-.42 1.586-2.029 2.538-3.601 2.141l-7.489-2.026a4.696 4.696 0 00-3.62.48 4.773 4.773 0 00-2.222 2.925 4.835 4.835 0 00.479 3.664 4.757 4.757 0 002.905 2.251 4.696 4.696 0 003.62-.48 4.773 4.773 0 002.222-2.924l.26-1.047a.25.25 0 00-.177-.302l-1.258-.34a.25.25 0 00-.307.18l-.256 1.025c-.426 1.602-2.064 2.556-3.651 2.127-1.582-.428-2.523-2.073-2.1-3.67.421-1.586 2.03-2.538 3.602-2.142l7.489 2.027a4.696 4.696 0 003.62-.48 4.773 4.773 0 002.222-2.925 4.835 4.835 0 00-.479-3.664 4.757 4.757 0 00-2.905-2.251 4.696 4.696 0 00-3.62.48 4.773 4.773 0 00-2.222 2.924l-.247.828a.25.25 0 00.174.312l1.259.34a.25.25 0 00.304-.169l.248-.827c.426-1.602 2.064-2.557 3.651-2.127 1.582.428 2.523 2.073 2.1 3.67z"
                  fill="url(#cd-main_svg__a)"
                  fill-rule="evenodd"
                />
                <defs>
                  <lineargradient
                    gradientUnits="userSpaceOnUse"
                    id="cd-main_svg__a"
                    x1="6.774"
                    x2="21.5"
                    y1="2"
                    y2="30"
                  >
                    <stop
                      stop-color="#62F91F"
                    />
                    <stop
                      offset="1"
                      stop-color="#45BD35"
                    />
                  </lineargradient>
                </defs>
              </svg>
            </span>
            common.purpose.cd.continuous
          </p>
          <p
            class="StyledProps--font StyledProps--main StyledProps--padding-bottom-small StyledProps--padding-left-large StyledProps--color StyledProps--color-white"
          >
            pipeline.featureRestriction.maxBuildsPerMonth100PercentLimit
          </p>
        </div>
        <button
          class="bp3-button Button--button StyledProps--font StyledProps--main btn Button--with-current-color Button--iconOnly Button--variation Button--variation-primary Button--size Button--small"
          type="button"
        >
          <span
            class="bp3-button-text"
          >
            common.explorePlans
          </span>
        </button>
      </div>
      <div
        class="Layout--vertical StyledProps--main StyledProps--padding-bottom-large"
      >
        <div
          class="Layout--vertical StyledProps--main StyledProps--padding-bottom-small"
        >
          <p
            class="StyledProps--font StyledProps--main StyledProps--color StyledProps--color-white StyledProps--font-weight-semi-bold StyledProps--padding-bottom-small"
            style="display: flex; align-items: center;"
          >
            <span
              class="bp3-icon StyledProps--main Text--icon StyledProps--padding-right-xsmall StyledProps--color StyledProps--color-white"
              data-icon="ci-main"
            >
              <svg
                height="16"
                viewBox="0 0 32 32"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  clip-rule="evenodd"
                  d="M29 16a13.02 13.02 0 01-.854 4.6.387.387 0 01-.637.13l-3.99-3.94a.534.534 0 01-.085-.628c.432-.827.649-1.773.649-2.762.056-3.222-2.6-5.878-5.879-5.878-1.04 0-1.994.26-2.823.706a.53.53 0 01-.623-.08l-3.551-3.504a.48.48 0 01.154-.794A12.79 12.79 0 0116 3c7.178 0 13 5.822 13 13zM8.476 5.42a.582.582 0 01.746.072l4.035 4.035c.179.178.194.46.054.67a5.829 5.829 0 00-.985 3.26 5.934 5.934 0 005.935 5.934 5.829 5.829 0 003.26-.985.531.531 0 01.67.054l4.367 4.368a.582.582 0 01.065.756C24.225 26.86 20.4 29 16 29 8.822 29 3 23.178 3 16c0-4.403 2.144-8.231 5.476-10.58zm9.785 11.71a3.932 3.932 0 003.956-3.956 3.932 3.932 0 00-3.956-3.957 3.932 3.932 0 00-3.957 3.957 3.932 3.932 0 003.957 3.956z"
                  fill="url(#ci-main_svg__a)"
                  fill-rule="evenodd"
                />
                <defs>
                  <lineargradient
                    gradientUnits="userSpaceOnUse"
                    id="ci-main_svg__a"
                    x1="-5.398"
                    x2="11.398"
                    y1="11.398"
                    y2="37.398"
                  >
                    <stop
                      stop-color="#73DFE7"
                    />
                    <stop
                      offset="1"
                      stop-color="#0095F7"
                    />
                  </lineargradient>
                </defs>
              </svg>
            </span>
            common.purpose.ci.continuous
          </p>
          <p
            class="StyledProps--font StyledProps--main StyledProps--padding-bottom-small StyledProps--padding-left-large StyledProps--color StyledProps--color-white"
          >
            common.feature.levelUp.limitMessage.description 
            Deployments
             common.feature.levelup.limitmessage.thismonth.
          </p>
        </div>
        <button
          class="bp3-button Button--button StyledProps--font StyledProps--main btn Button--with-current-color Button--iconOnly Button--variation Button--variation-primary Button--size Button--small"
          type="button"
        >
          <span
            class="bp3-button-text"
          >
            common.explorePlans
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;
